storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1268 @@
+			return -EIO;
+		}
+	}
+	else {
+		TRACE("HS mode not supported");
+		if (sdcard_configClockAndPower(host, SD_FREQ_25M) < 0) {
+			return -EIO;
+		}
+	}
+
+	usleep(10);
+
+	/* Perform a final transaction to check if data transfer is working (we only care about retcode this time) */
+	if (sdio_cmdSendEx(host, SDIO_ACMD13_SD_STATUS, 0, NULL, 0, NULL) < 0) {
+		LOG_ERROR("bus speed change failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+
+static int sdcard_startEventISR(sdcard_hostData_t *host, int interruptNum)
+{
+	if (mutexCreate(&host->eventLock) < 0) {
+		return -ENOMEM;
+	}
+
+	if (condCreate(&host->eventCond) < 0) {
+		resourceDestroy(host->eventLock);
+		return -ENOMEM;
+	}
+
+	interrupt(interruptNum, sdhost_isr, host, host->eventCond, &host->isrHandle);
+
+	return 0;
+}
+
+
+static void _sdcard_free(sdcard_hostData_t *host)
+{
+	TRACE("freeing resources");
+	sdhost_reset(host, CLOCK_CONTROL_RESET_ALL);
+	host->sdioInitialized = 0;
+
+	if (host->cmdLock != 0) {
+		resourceDestroy(host->cmdLock);
+		host->cmdLock = -1;
+	}
+
+	if (host->dmaBuffer != NULL) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		host->dmaBufferPhys = (addr_t)NULL;
+	}
+
+	*(host->base + SDHOST_REG_INTR_STATUS_ENABLE) = 0;
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) = 0;
+}
+
+
+void sdcard_free(unsigned int slot)
+{
+	GET_HOST_OR_RETURN();
+	if (host->sdioInitialized == 0) {
+		return;
+	}
+
+	mutexLock(host->eventLock);
+	_sdcard_free(host);
+	mutexUnlock(host->eventLock);
+}
+
+
+int sdcard_initHost(unsigned int slot)
+{
+	void *ptr;
+	sdio_platformInfo_t info;
+
+	if (!presenceEvents.initialized) {
+		if (mutexCreate(&presenceEvents.lock) < 0) {
+			return -ENOMEM;
+		}
+
+		if (condCreate(&presenceEvents.cond) < 0) {
+			resourceDestroy(presenceEvents.lock);
+			return -ENOMEM;
+		}
+
+		presenceEvents.initialized = 1;
+	}
+
+	GET_HOST_OR_RETURN(-ENOENT);
+
+	if (host->sdioInitialized) {
+		return (initializedHosts < PLATFORM_SDIO_N_HOSTS) ?
+			(PLATFORM_SDIO_N_HOSTS - initializedHosts) :
+			0;
+	}
+
+	host->cmdLock = 0;
+	host->eventCond = 0;
+	host->eventLock = 0;
+
+	/* Perform platform-specific configuration */
+	if (sdio_platformConfigure(slot, &info) < 0) {
+		return -EIO;
+	}
+
+	host->refclkFrequency = info.refclkFrequency;
+	host->isCDPinSupported = info.isCDPinSupported;
+	host->isWPPinSupported = info.isWPPinSupported;
+
+	/* Map register bank into our virtual memory */
+	ptr = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, info.regBankPhys);
+	if (ptr == MAP_FAILED) {
+		_sdcard_free(host);
+		return -EIO;
+	}
+
+	host->base = ptr;
+	if (sdhost_allocDMA(host) < 0) {
+		_sdcard_free(host);
+		return -ENOMEM;
+	}
+
+	if (mutexCreate(&host->cmdLock) < 0) {
+		_sdcard_free(host);
+		return -ENOMEM;
+	}
+
+	if (sdhost_reset(host, CLOCK_CONTROL_RESET_ALL) < 0) {
+		_sdcard_free(host);
+		return -EIO;
+	}
+
+	sdcard_configClockAndPower(host, SD_FREQ_INITIAL);
+
+	if (!host->isCDPinSupported) {
```

Comment:
MISRA
```suggestion
	if (host->isCDPinSupported == 0) {
```

Reply:
I changed `host->isCDPinSupported` to `bool` so this code can stay as is.